File: server/src/routes/item.js
const express = require('express');
const router = express.Router();

const Item = require('../models/itemModel');

// Create item request
router.post('/create', async (req, res) => {

    const { item_name, username, password, uri, notes } = req.body;

    // Check if item_name, username, or password is empty, uri and notes 
    // are not checked since they are allowed to be empty.
    if (!item_name || !username || !password) { 
        return res.status(400).json({ message: "Empty item_name/username/password field(s)" });
    }

    const success = Item.create(req.session.accountId, item_name, username, password, uri, notes);
    if (success) {
        return res.status(200).json({ message: "Item created" });
    }

    return res.status(400).json({ message: "Item not created" });
});

// Get item associated with the logged in user request
router.get('/get/:itemId', async (req, res) => {
    // Get item with id '1' using "http://localhost:8080/api/item/get/1"
    const { itemId } = req.params;
    const result = Item.get(req.session.accountId, itemId);
    if (result) {
        return res.status(200).json(result);
    }

    return res.status(500).json({ message: "Failed to get item" });
});

// Get all items associated with the logged in user request
router.get('/getAll', async (req, res) => {

    const result = Item.getAll(req.session.accountId);
    if (result) {
        return res.status(200).json(result);
    }

    return res.status(500).json({ message: "Failed to get items" });
});

router.post("/update", async (req, res) => {
    const { itemId, data } = req.body;

    // Check if itemId is empty
    if (!itemId) {
        return res.status(400).json({ message: "Empty itemId field" });
    }
    // Check if data is empty
    if (!data || Object.keys(data).length == 0) {
        return res.status(400).json({ message: "Empty data field" });
    }

    const success = Item.update(req.session.accountId, itemId, data);
    if (success) {
        return res.status(200).json({ message: "Item updated" });
    }

    return res.status(400).json({ message: "Item not updated" });
})

// Delete item request
router.post("/delete", async (req, res) => {

    const { itemId } = req.body;

    // Check if itemId is empty
    if (!itemId) {
        return res.status(400).json({ message: "Empty itemId field" });
    }

    const success = Item.delete(req.session.accountId, itemId);
    if (success) {
        return res.status(200).json({ message: "Item removed" });
    }

    return res.status(400).json({ message: "Item not removed" });
});

module.exports = router

File: application/app/src/main/java/communication/Item.kt
package communication

import com.application.keykeeper.BuildConfig
import com.github.kittinunf.fuel.core.extensions.jsonBody
import communication.structure.DatabaseItem
import communication.structure.ServerMessage
import org.json.JSONObject
import structure.CredentialsItem

object Item {

    private const val notLoggedInMessage: String = "User is not logged in."

    private fun jsonItemData(item: CredentialsItem): JSONObject {
        // Create object containing data to be sent to the server.
        val itemData = JSONObject()

        // TODO: Encrypt all information that is sent to the server.
        itemData.put("item_name", item.label)
        itemData.put("username", item.username)
        itemData.put("password", item.password)
        itemData.put("uri", item.uri)
        itemData.put("notes", item.notes)

        return itemData
    }

    private fun jsonDeleteData(item: CredentialsItem): JSONObject {
        // Create object containing data to be sent to the server.
        val itemData = JSONObject()
        itemData.put("itemId", item.id)
        return itemData
    }

    private fun jsonUpdateData(item: CredentialsItem): JSONObject {
        // Create object containing data to be sent to the server.
        val serverData = JSONObject()

        serverData.put("itemId", item.id)
        serverData.put("data", jsonItemData(item))

        return serverData
    }

    private fun matchResponseCodeToMessage(responseCode: Int, statusOkMessage: String?): String {
        return when(responseCode) {
            200 -> statusOkMessage ?: "No message."
            400 -> "Bad input data. Try again."
            403 -> "Cannot execute request when not logged in."
            else -> "Something went wrong when communicating with the server. Try again later."
        }
    }

    private fun sendAuthenticatedPostRequestWithServerMessageResponse(url: String, jsonData: JSONObject, callback: (successful: Boolean, message: String) -> Unit) {
        Account.sendAuthenticatedPostRequest(url)
            .jsonBody(jsonData.toString())
            .responseObject(ServerMessage.getDeserializer()) { _, response, result ->
                val (serverResponse, _) = result

                callback.invoke(
                    response.statusCode == 200,
                    matchResponseCodeToMessage(response.statusCode, serverResponse?.message)
                )
            }
    }

    fun sendCreateItemRequest(item: CredentialsItem, callback: (successful: Boolean, message: String) -> Unit) {
        val jsonPostData = this.jsonItemData(item)

        if(!Account.isLoggedIn()) {
            callback.invoke(false, notLoggedInMessage)
        } else {
            sendAuthenticatedPostRequestWithServerMessageResponse(
                 BuildConfig.SERVER_URL + "api/item/create",
                jsonPostData,
                callback
            )
        }
    }

    fun sendGetItemsRequest(callback: (successful: Boolean, message: String, items: Array<DatabaseItem>) -> Unit) {
        if(!Account.isLoggedIn()) {
            callback.invoke(false, notLoggedInMessage, arrayOf())
        } else {
            Account.sendAuthenticatedGetRequest(BuildConfig.SERVER_URL + "api/item/getAll")
                .responseObject(DatabaseItem.getArrayDeserializer()) { _, response, result ->
                    val (serverResponse, _) = result

                    callback.invoke(
                        response.statusCode == 200,
                        matchResponseCodeToMessage(response.statusCode, null),
                        serverResponse ?: arrayOf()
                    )
                }
        }
    }

    fun sendDeleteItemRequest(item: CredentialsItem, callback: (successful: Boolean, message: String) -> Unit) {
        val jsonPostData = this.jsonDeleteData(item)

        if(!Account.isLoggedIn()) {
            callback.invoke(false, notLoggedInMessage)
        } else {
            sendAuthenticatedPostRequestWithServerMessageResponse(
                BuildConfig.SERVER_URL + "api/item/delete",
                jsonPostData,
                callback
            )
        }
    }

    fun sendUpdateItemRequest(item: CredentialsItem, callback: (successful: Boolean, message: String) -> Unit) {
        val jsonPostData = this.jsonUpdateData(item)

        if(!Account.isLoggedIn()) {
            callback.invoke(false, notLoggedInMessage)
        } else {
            sendAuthenticatedPostRequestWithServerMessageResponse(
                BuildConfig.SERVER_URL + "api/item/update",
                jsonPostData,
                callback
            )
        }
    }
}